roqoqo/src/circuitdag.rs

```diff
@@ -0,0 +1,985 @@
+    }
+
+    /// Returns a reference to the HashMap where a key is composed by the name and the size
+    /// of the classical register and its value represents the last node that involves that
+    /// register.
+    ///
+    pub fn last_operation_involving_classical(&self) -> &HashMap<(String, usize), usize> {
+        &self.last_operation_involving_classical
+    }
+
+    /// Given a NodeIndex, returns a reference to the Operation contained in the node of
+    /// the CircuitDag.
+    ///
+    pub fn get(&self, node: NodeIndex<usize>) -> Option<&Operation> {
+        self.graph.node_weight(node.into())
+    }
+}
+
+/// Creates a new CircuitDag from a given Circuit.
+///
+impl From<Circuit> for CircuitDag {
+    fn from(circuit: Circuit) -> Self {
+        let mut new_dag = CircuitDag {
+            graph: Graph::<Operation, (), Directed, usize>::with_capacity(
+                circuit.len(),
+                circuit.operations().len(),
+            ),
+            commuting_operations: Vec::<NodeIndex<usize>>::new(),
+            first_parallel_block: HashSet::<NodeIndex<usize>>::new(),
+            last_parallel_block: HashSet::<NodeIndex<usize>>::new(),
+            first_all: Option::<NodeIndex<usize>>::None,
+            last_all: Option::<NodeIndex<usize>>::None,
+            first_operation_involving_qubit: HashMap::<usize, NodeIndex<usize>>::new(),
+            last_operation_involving_qubit: HashMap::<usize, NodeIndex<usize>>::new(),
+            first_operation_involving_classical: HashMap::<(String, usize), NodeIndex<usize>>::new(
+            ),
+            last_operation_involving_classical: HashMap::<(String, usize), NodeIndex<usize>>::new(),
+            _roqoqo_version: RoqoqoVersion,
+        };
+
+        for operation in circuit.operations() {
+            new_dag.add_to_back(operation.clone());
+        }
+
+        new_dag
+    }
+}
+
+/// Creates a new Circuit from a given CircuitDag.
+///
+impl From<CircuitDag> for Circuit {
+    fn from(dag: CircuitDag) -> Circuit {
+        let mut circuit: Circuit = Circuit::new();
+
+        match toposort(&dag.graph, None) {
+            Ok(order) => {
+                for i in order {
+                    circuit.add_operation(dag.graph.node_weight(i).unwrap().clone());
+                }
+            }
+            Err(_) => {
+                println!("Error: graph not acyclic");
+            }
+        }
+
+        circuit
+    }
+}
+
+impl<'a> Iterator for ParallelBlocks<'a> {
+    type Item = Vec<(NodeIndex<usize>, Operation)>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // First case
+        if self.parallel_block.is_empty() && self.already_executed.is_empty() {
+            for node in &self.dag.first_parallel_block {
+                let op: Operation = self.dag.graph.node_weight((*node).into()).unwrap().clone();
+                self.parallel_block.push((*node, op));
+            }
+            return Some(self.parallel_block.clone());
+        }
```

Comment:
Here we probably want to Change the interface to 
either return a Vec<(usize, &Operation)> or even just a Vec<usize> and let the Rust users get the Operation reference with a getter function.
Using clone on the operations is expensive and the users can do it if they need and we simplified all other returns of indices to just `usize` as well.